mediatimestamp/immutable/timestamp.py

```diff
@@ -71,36 +146,94 @@ def __mediatimerange__(self) -> "TimeRange":
         unix_time = time.time()
         return cls.from_unix(int(unix_time), int(unix_time*cls.MAX_NANOSEC) - int(unix_time)*cls.MAX_NANOSEC)
 
     @classmethod
-    def from_timeoffset(cls, toff: TimeOffsetConstructionType) -> "Timestamp":
-        toff = mediatimeoffset(toff)
+    @deprecated(version="4.0.0",
+                reason="This method is deprecated. TimeOffset has been merged into Timestamp.")
+    def from_timeoffset(cls, toff: TimestampConstructionType) -> "Timestamp":
+        """Legacy method that converted a TimeOffset to a Timestamp"""
+        toff = mediatimestamp(toff)
         return cls(sec=toff.sec, ns=toff.ns, sign=toff.sign)
 
     @classmethod
-    def from_sec_frac(cls, ts_str: str) -> "Timestamp":
-        return cast(Timestamp, super(Timestamp, cls).from_sec_frac(ts_str))
+    def get_interval_fraction(cls,
+                              rate_num: RationalTypes,
+                              rate_den: RationalTypes = 1,
+                              factor: int = 1) -> "Timestamp":
+        if rate_num <= 0 or rate_den <= 0:
+            raise TsValueError("invalid rate")
+        if factor < 1:
+            raise TsValueError("invalid interval factor")
+
+        rate = Fraction(rate_num, rate_den)
+        ns = int((cls.MAX_NANOSEC * rate.denominator) // (rate.numerator * factor))
+        return cls(ns=ns)
+
+    @classmethod
+    def from_millisec(cls, millisec: int) -> "Timestamp":
+        ns = millisec * 1000**2
+        return cls(ns=ns)
+
+    @classmethod
+    def from_microsec(cls, microsec: int) -> "Timestamp":
+        ns = microsec * 1000
+        return cls(ns=ns)
 
     @classmethod
     def from_nanosec(cls, nanosec: int) -> "Timestamp":
-        return cast(Timestamp, super(Timestamp, cls).from_nanosec(nanosec))
+        return cls(ns=nanosec)
+
+    @classmethod
+    def from_sec_frac(cls, toff_str: str) -> "Timestamp":
+        sec_frac = toff_str.split(".")
+        if len(sec_frac) != 1 and len(sec_frac) != 2:
+            raise TsValueError("invalid second.fraction format")
+        sec = int(sec_frac[0])
+        sign = 1
+        if sec_frac[0].startswith("-"):
+            sign = -1
+            sec = abs(sec)
+        ns = 0
+        if len(sec_frac) > 1:
+            ns = _parse_seconds_fraction(sec_frac[1])
+        return cls(sec=sec, ns=ns, sign=sign)
 
     @classmethod
     def from_tai_sec_frac(cls, ts_str: str) -> "Timestamp":
         return cls.from_sec_frac(ts_str)
 
     @classmethod
-    def from_sec_nsec(cls, ts_str: str) -> "Timestamp":
-        return cast(Timestamp, super(Timestamp, cls).from_sec_nsec(ts_str))
+    def from_sec_nsec(cls, toff_str: str) -> "Timestamp":
+        sec_frac = toff_str.split(":")
+        if len(sec_frac) != 1 and len(sec_frac) != 2:
+            raise TsValueError("invalid second:nanosecond format")
+        sec = int(sec_frac[0])
+        sign = 1
+        if sec_frac[0].startswith("-"):
+            sign = -1
+            sec = abs(sec)
```

Comment:
I suppose you could use `sec_frac[0]` directly since the constructor supports a signed `sec` value, but maybe this is clearer

Reply:
The nanoseconds would need the sign as well. The `sign` arg in the constructor is a little unusual and maybe we should look to remove it at the next breaking change / deprecation completion.